components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,322 @@
+
+        if self.exists().unwrap() {
+            Ok(())
+        } else {
+            match self.client.create_bucket(&request).sync() {
+                Ok(_response) => Ok(()),
+                Err(e) => {
+                    debug!("{:?}", e);
+                    Err(Error::CreateBucketError(e))
+                }
+            }
+        }
+    }
+
+    pub fn upload(
+        &self,
+        hart: &PathBuf,
+        ident: &PackageIdent,
+        target: &PackageTarget,
+    ) -> Result<()> {
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        let file = File::open(hart)?;
+        let size = file.metadata().unwrap().len() as usize;
+        let min_limit: usize = 49152 * 1024;
+        if size < min_limit {
+            self.single_upload(key, file)
+        } else {
+            self.multipart_upload(key, file)
+        }
+    }
+
+    pub fn download(
+        &self,
+        loc: &PathBuf,
+        ident: &PackageIdent,
+        target: &PackageTarget,
+    ) -> Result<()> {
+        let mut request = GetObjectRequest::default();
+        let key = self.s3_key(ident, target).to_string_lossy().into_owned();
+        request.bucket = self.bucket.to_owned();
+        request.key = key;
+
+        let payload = self.client.get_object(&request).sync();
+        let body = match payload {
+            Ok(response) => response.body,
+            Err(e) => {
+                return Err(Error::PackageDownload(e));
+            }
+        };
+
+        let file = body.expect("Downloaded pkg archive empty!").concat2();
+        let _result = DepotUtil::write_archive(&loc, file.wait().unwrap());
+        Ok(())
+    }
+
+    fn single_upload(&self, key: String, hart: File) -> Result<()> {
+        let mut reader = BufReader::new(hart);
+        let mut object: Vec<u8> = Vec::new();
+        let bucket = self.bucket.clone();
+        let _result = reader.read_to_end(&mut object);
+
+        let mut request = PutObjectRequest::default();
+        request.key = key;
+        request.bucket = bucket;
+        request.body = Some(object);
+
+        match self.client.put_object(&request).sync() {
+            Ok(_) => Ok(()),
+            Err(e) => {
+                debug!("{:?}", e);
+                Err(Error::PackageUpload(e))
+            }
+        }
+    }
+
+    fn multipart_upload(&self, key: String, hart: File) -> Result<()> {
+        let mut p: Vec<CompletedPart> = Vec::new();
+        let mut mprequest = CreateMultipartUploadRequest::default();
+        mprequest.key = key.clone();
+        mprequest.bucket = self.bucket.clone();
+
+        match self.client.create_multipart_upload(&mprequest).sync() {
+            Ok(output) => {
+                let mut request = UploadPartRequest::default();
+                request.key = key.clone();
+                request.bucket = self.bucket.clone();
+                request.upload_id = output.upload_id.clone().unwrap();
```

Comment:
Unwrap ok?

Reply:
Yep this should be good, the rusoto library has this struct parameter listed with the Option type but AWS multipart upload requests should always return an upload id in the case of an OK response. This _should_ be ok I believe.

Reply:
OK (may be good to also add an 'unwrap safe' comment at the end of the line...)